### PR TITLE
[layer] Share gradient weights for all the layers @open sesame 12/07 09:04

### LINKED ISSUE #766: Gradients don't need own memory
Gradients for the weights are just a temporary memory, as they are only useful till the gradient is applied to the layer or when the gradient is used for debugging.
So, gradients of all the layers can share the same tensor with the same memory. This memory would be the maximum sized gradient needed by a layer. 
With this, the gradients memory for a model is just reduced to the gradients memory of layer, reducing the peak memory of the model by a lot.

### PR DESCRIPTION
**Pull request by kparichay**
**Commit 1**: [weight/var_grad] Make internal variable as shared_ptr
Internal variables in weights/var_grad, namely, the variable and gradient itself
    are changed to shared_ptr so that weights can be shared without worrying about
    shallow copies.
    
Also changed the copy constructor to not create new Tensor as copy constructor
    of weight will get called and its unnecessary + unintentional overhead.
    As weight is just wrapper over tensor, their copy constructors should follow
    same behavior as tensor which is to not create new memory.
    Added clone as an alternative to create new copy of a given weight.
    
**Commit 2**: [manager] Added nntrainer manager for weights
Added manager to manage all the allocated weights
    This patch also adds manager to the model and passes manager to the
    initialize which allows weights to be added to the manager.

**Commit 3**: [weight] Updated weights to be vector
Updated weights of layer to be vector than a shared_ptr array
    This is for easier management and updating weight internally when
    gradient will share the memory

**Commit 4**: [layers/manager] Register weights with manager
All the weights of the layer are now registered with manager
    Manager allocates memory for these weights and in future
    handle their updates etc

**Commit 5**: [manager] Share gradient memory for all layer
This patch allows sharing the memory for gradient across all the layers
    The maximum size of the gradient is allocated and all layers have unique tensors
    which internally points to this tensor.
    
This optimization feature can be disabled for a model (as done with automated models unittest)
    
Manager is also moved to nntrainer/tensor as manager is managing all the weights (tensors) and will
    in future manage all the inputs/outputs.
    If the functionality of manager is extended, then it can be appropriately moved.
    
See also #774
Resolves #766
    
**Self evaluation:**
    1. Build test: [x]Passed [ ]Failed [ ]Skipped
    2. Run test: [x]Passed [ ]Failed [ ]Skipped
    
Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM. Currently manager only manage the weight. but I think it should work same way for tensors in graph. and every layer include this manager so.. add tensor vector for input tensor and manage this and handle sharing between layers. I guess it could be sapearte PR.

**Comment by kparichay** [comment]
> Good work. LGTM. Currently manager only manage the weight. but I think it should work same way for tensors in graph. and every layer include this manager so.. add tensor vector for input tensor and manage this and handle sharing between layers. I guess it could be sapearte PR.

Yes, that's the big plan. Manager will manage all the tensors and its related optimizations. Input, output and temporary allocations of the tensor, all will go with manager in a separate PR.

File: test/unittest/unittest_nntrainer_models.cpp [hunk 1]
```diff
@@ -87,8 +87,8 @@ class NodeWatcher {
     node.layer->setTrainable(true);
 
     for (unsigned int i = 0; i < num_weights; ++i) {
-      const nntrainer::Weight &w = node.layer->weightAt(i);
-      expected_weights.push_back(w);
+      nntrainer::Weight &w = node.layer->weightAt(i);
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_models.cpp:90]: I guess this should remain const and change signature of `weight::clone const ()`?

File: nntrainer/manager.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+/**
+ * @brief Helper func for weight creation which are tracked by nntrainer
+ *
+ * @retval create weight
+ */
+template <typename... Args>
+Weight createWeight(Manager &manager, Args... args) {
+  Weight w = Weight(args...);
```
Comment (zhoonit) [inline nntrainer/manager.h]: `std::forward<Args>(args)...`  to do perfect forwarding?

Comment (zhoonit) [reply nntrainer/manager.h]: Just an idea though, How about migrate this to one of weight constructor?

File: nntrainer/manager.h [hunk 3]
```diff
@@ -52,32 +55,55 @@ class Manager {
...
- * @retval create weight
- */
-template <typename... Args>
-Weight createWeight(Manager &manager, Args... args) {
-  Weight w = Weight(args...);
-  manager.trackWeight(w);
-  return w;
-}
+// /**
+//  * @brief Helper func for weight creation which are tracked by nntrainer
```
Comment (zhoonit) [inline nntrainer/manager.h]: Maybe this can be deleted??

File: nntrainer/tensor/weight.cpp [hunk 4]
```diff
@@ -23,10 +23,12 @@ Weight::Weight(const TensorDim &dim, const WeightInitializer init, bool train,
   if (initializer == WeightInitializer::WEIGHT_UNKNOWN)
     throw std::invalid_argument("Weight initializer unknown");
 
-  initializeWeight();
+  // initialize();
```
Comment (zhoonit) [inline nntrainer/tensor/weight.cpp]: Same

File: nntrainer/tensor/manager.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+#include <vector>
+
+#include <manager.h>
+
+namespace nntrainer {
+
+/**
+ * @brief     Add weight to be tracked and updated with nntrainer
+ */
+void Manager::trackWeight(std::reference_wrapper<Weight> w) {
```
Comment (zhoonit) [inline nntrainer/tensor/manager.cpp:25]: I wonder if we can't do `Weight &w` for the signature

**Review by zhoonit — APPROVED** [review]
Overall LGTM Great work!

File: nntrainer/manager.h (hunk 2 above)
Comment (kparichay) [reply nntrainer/manager.h]: This is later removed as this does not work.

File: nntrainer/manager.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/manager.h]: done.

File: nntrainer/tensor/weight.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/tensor/weight.cpp]: done.

File: test/unittest/unittest_nntrainer_models.cpp (hunk 1 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_models.cpp]: done.

**Pull request merged**